src/cudaautotune/autotuning/tuner.py

```diff
@@ -0,0 +1,149 @@
+        help='run the process multiple times before autotuning. Default = False')
+parser.add_argument('--verbose', '-v', action='count', default=0)
+
+args = parser.parse_args()
+
+tunables = {}
+tunables_list = []
+lbounds = []
+ubounds = []
+steps = []
+
+tunables_file = args.tunables[0]
+for n, line in enumerate(tunables_file.readlines()):
+    parameters = line.strip('\n').split(',')
+
+    if args.verbose > 1:
+        print(parameters)
+
+    tunables[parameters[0]] = n
+    tunables_list.append(parameters[0])
+    lbounds.append(int(parameters[1]))
+    ubounds.append(int(parameters[2]))
+    steps.append(int(parameters[3]))
+
+tunables_file.close()
+ 
+configurations = []
+ranges = [range(l, u + 1, s) for l, u, s in zip(lbounds, ubounds, steps)]
+length = 1
+for r in ranges:
+    r = list(r)
+    length = length * len(r)
+    random.shuffle(r)
+    configurations.append(r)
+
+if args.verbose:
+    print("number of configurations is " + str(length))
+
+# Heating up the GPU before tuning
+process_path = args.process[0].resolve()
+
+if args.preheat:
+    if args.verbose:
+        print('Preheating started')
+    cmd = [process_path, "--runForMinutes", "1", "--numberOfThreads", "12", "--numberOfStreams", "12"]
+    if args.verbose > 1:
+        print('Preheat command:', cmd)
+    subprocess.run(cmd)
+    if args.verbose:
+        print('Preheating finished')
+
+# Tuning
+configurations = list(itertools.product(*configurations))
+random.shuffle(configurations)
+results_file = args.results[0]
+kernels_config_path = args.configurations[0]
+kernels = tunables_list[:-2]
+for config in configurations:
+    if args.verbose:
+        print('Configuration:', config)
+
+    # TODO write only when the parameter change
+    for kernel in kernels:
+        file = open(kernels_config_path / kernel, 'w')
+        file.write(str(config[tunables[kernel]]))
+        file.close()
+    
+    # Validation
+    validation = ""
+    cmd = [process_path,
+            "--numberOfThreads", "12",
+            "--numberOfStreams", "12",
+            "--validation"]
+    if args.verbose > 1:
+        print('Validation command:', cmd)
+    process = subprocess.run(cmd, encoding='UTF-8', capture_output=True)
+    if (process.returncode == 0):
+        if args.verbose > 1:
+            print('Validation output:', process.stdout)
+        if process.stdout.find("passed validation"):
+            validation = "PASSED"
+        elif process.stdout.find("failed validation"):
+            validation = "FAILED"
+        else:
+            validation = "ERROR"
+    else:
+        print(process.stdout)
+        validation = "ERROR"
+    if args.verbose:
+        print('Validation result:', validation)
+        
+    # Benchmarking
+    time = "NaN"
+    throughput = "NaN"
+    cpu_efficiency = "NaN"
+    status = ""
+
+    cpu_threads = config[tunables["cpu_threads"]]
+    gpu_streams = cpu_threads + config[tunables["gpu_streams"]]
```

Comment:
It is not clear to me why the number of concurrent events (`--numberOfStreams` below) is set to "number of CPU threads" + something named "GPU streams". Could you clarify what is the intended behavior here wrt. CPU threads and concurrent events?